extensions/amp-ad-network-adsense-impl/0.1/amp-ad-network-adsense-impl.js

```diff
@@ -110,14 +138,29 @@ export class AmpAdNetworkAdsenseImpl extends AmpA4A {
       {name: 'u_w', value: screen ? screen.width : null},
       {name: 'vis', value: visibilityStateCodes[visibilityState] || '0'},
       {name: 'wgl', value: global['WebGLRenderingContext'] ? '1' : '0'},
-    ], []);
+    ];
+
+    if (sharedStateParams.prevFmts) {
+      paramList.push({name: 'prev_fmts', value: sharedStateParams.prevFmts});
+    }
+
+    return googleAdUrl(
+        this, ADSENSE_BASE_URL, startTime, slotIdNumber, paramList, []);
   }
 
   /** @override */
   extractCreativeAndSignature(responseText, responseHeaders) {
     return extractGoogleAdCreativeAndSignature(responseText, responseHeaders);
   }
 
+  /** @override */
+  unlayoutCallback() {
+    super.unlayoutCallback();
+    if (this.adk_) {
+      sharedState.removeSlot(this.adk_);
```

Comment:
Are we guaranteed adk is unique across slots on the same page?   You could concat with slotId to be safe otherwise the removeSlot would potentially cause multiple slots to be removed